Tools/AP_Periph/esc_apd_telem.cpp

```diff
@@ -0,0 +1,103 @@
+/*
+  ESC Telemetry for Hobbywing Pro 80A HV ESC. This will be
+  incorporated into a broader ESC telemetry library in ArduPilot
+  master in the future
+
+  This protocol only allows for one ESC per UART RX line, so using a
+  CAN node per ESC works well.
+ */
+#include "esc_apd_telem.h"
+#include <AP_HAL/utility/sparse-endian.h>
+#include <AP_Math/crc.h>
+#include <AP_Math/definitions.h>
+#include <string.h>
+
+#ifdef HAL_PERIPH_ENABLE_ESC_APD
+
+extern const AP_HAL::HAL& hal;
+
+#define TELEM_HEADER 0x9B
+#define TELEM_LEN    0x16
+
+ESC_APD_Telem::ESC_APD_Telem(void) { }
+
+void ESC_APD_Telem::init(AP_HAL::UARTDriver *_uart, float num_poles) {
+    uart = _uart;
+    uart->begin(115200);
+
+    pole_count = num_poles;
+}
+
+bool ESC_APD_Telem::update() {
+    uint32_t n = uart->available();
+    if (n == 0) {
+        return false;
+    }
+
+    // don't read too much in one loop to prevent too high CPU load
+    if (n > 500) {
+        n = 500;
+        uart->discard_input();
+    }
+
+    bool ret = false;
+
+    while (n--) {
+        uint8_t b = uart->read();
+        received.bytes[len++] = b;
+
+        // check the packet size first
+        if ((size_t)len == sizeof(received.packet)) {
+            // we have a full packet, check the stop byte
+            if (received.packet.stop == 65535) {
+                // valid stop byte, check the CRC
+                if (crc_fletcher16(received.bytes, 18) == received.packet.checksum) {
+                    // valid packet, copy the data we need and reset length
+                    ret = true;
+                    decoded.voltage = le16toh(received.packet.voltage) * 1e-2f;
+                    decoded.temperature = convert_temperature(le16toh(received.packet.temperature));
+                    decoded.current = le16toh(received.packet.bus_current) * (1 / 12.5);
+                    decoded.rpm = le32toh(received.packet.erpm) / pole_count;
+                    decoded.power_rating_pct = le16toh(received.packet.motor_duty) / 100;
```

Comment:
* 1e-2f in one place, and "/ 100" in another? You know how I feel about that "/" symbol... :)